### PR TITLE
Virtual Memory overhaul, part 1

### PR DESCRIPTION
**Pull request by sethpugsley**
This is the first step in implementing the ideas presented in Issue #49 

It replaces the va_to_pa() function with a new class, called VirtualMemory.  The VirtualMemory class has a function, va_to_pa(), which offers the same functionality as the old function, but is much simpler and faster, and will be easier to integrate with future virtual memory system changes like TLBs and page walkers.  The capacity, page size, and number of levels in the page table are all configurable.  It also uses its own random number function to shuffle the physical address space, so as to not interfere with any other RNG systems.

The new class also allows you to get the address of the page table entry for a given virtual address in each level of the page table.  This is the part of the code that I'm less sure of, and would appreciate some eyeballs on.  This code isn't used by anything in ChampSim yet, but would need to work correctly before we can add page walkers.

### CODE REVIEW

File: inc/vmem.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
+
+#include <iostream>
+#include <deque>
+#include <map>
+
+#define VMEM_RAND_FACTOR 91827349653
```
Comment (ngober) [inline inc/vmem.h:6]: I don't understand the utility of this. Why is this needed?

File: src/main.cc [hunk 2]
```diff
@@ -783,6 +580,9 @@ int main(int argc, char** argv)
     uncore.LLC.llc_initialize_replacement();
     uncore.LLC.llc_prefetcher_initialize();
 
+    // initialize virtual memory system
+    vmem = new VirtualMemory(NUM_CPUS, 8589934592, 4096, 4, 1);
```
Comment (ngober) [inline src/main.cc:584]: Here, we have a hard-coded constant as the RNG seed. Is this to achieve deterministic shuffling? If so, why is that desirable?

File: src/vmem.cc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      vpage_to_ppage_map[cpu_num][vpage] = get_next_free_ppage();
+    }
+  
+  return (((vpage_to_ppage_map[cpu_num][vpage])<<log2_page_size)+voffset);
+}
+
+uint64_t VirtualMemory::get_pte_pa(uint32_t cpu_num, uint64_t vaddr, uint32_t level)
+{
+  uint64_t vpage = vaddr>>log2_page_size;
+  uint64_t pte_offset = vpage&511;
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/vmem.cc:113]: 512 is the number of PTEs per 4KB page in x86-64.  Each PTE is 8 bytes, hence the (pte_offset*8) later on.  In x86-64, even if you're using 2MB or 1GB pages, your page table is still composed of 4KB pages, as far as I know.

File: src/main.cc [hunk 4]
```diff
@@ -19,6 +19,8 @@ uint64_t warmup_instructions     = 1000000,
 
 time_t start_time;
 
+VirtualMemory* vmem;
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/main.cc:22]: I have no problem changing it to work that way.

File: src/vmem.cc [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+VirtualMemory::VirtualMemory(uint32_t number_of_cpus, uint64_t capacity, uint64_t pg_size, uint32_t page_table_levels, uint64_t random_seed)
+{
+  num_cpus = number_of_cpus;
+  pt_levels = page_table_levels;
+
+  // calculate number of physical pages
+  page_size = pg_size;
+  if(capacity%page_size != 0)
+    {
+      std::cout << "VirtualMemory initialization error: memory capacity must be a multiple of page size!" << std::endl;
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/vmem.cc]: I think we should have a comment explaining that all of these variables are counting bytes to make it more clear, but I don't think there's a problem with having an opportunity for illegal values during initialization.  I care more about having the call to the constructor clearly show the total physical memory capacity, so the user doesn't have to do math in their head to make sure the capacity is what they think they set it to.

File: src/vmem.cc [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      uint64_t swap_temp = ppage_free_list[i];
+      ppage_free_list[i] = ppage_free_list[swap_target];
+      ppage_free_list[swap_target] = swap_temp;
+    }
+  std::cout << "done" << std::endl << std::endl;
+
+  // initialize V to P page map tables
+  vpage_to_ppage_map = new std::map<uint64_t, uint64_t>[num_cpus];
+
+  // initialize per-process page tables
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/vmem.cc]: We have the code pattern "variable[cpu][...]" elsewhere in the code, so I went with that.  We don't have the code pattern of using pairs to index a map anywhere else in the code.  I also believe that vpage_to_ppage_map[cpu][vpage] is more illustrative of what goes on in a real virtual memory system.  Each process has its own completely separate page table.

File: src/vmem.cc [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+VirtualMemory::~VirtualMemory()
+{
+}
+
+uint64_t VirtualMemory::get_next_free_ppage()
+{
+  if(ppage_free_list.empty())
+    {
+      // ran out of physical pages to allocate, so throw error and exit
+      std::cout << "VirtualMemory error: ran out of physical pages to allocate!  Try a larger memory size." << std::endl;
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/vmem.cc:82]: I would rather leave this for later, if anyone ever requests this functionality.  In order to do it right, we'd have to also do TLB shootdowns and simulate disk access latency.  I think it would be a lot of work that nobody would even appreciate, because so far nobody's using ChampSim for page swapping studies.  I believe the current solution is preferable right now.

File: src/vmem.cc [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+  // initialize per-process page tables
+  page_table = new std::map<uint64_t, uint64_t>*[num_cpus];
+  for(uint32_t i=0; i<num_cpus; i++)
+    {
+      page_table[i] = new std::map<uint64_t, uint64_t>[pt_levels];
+    }
+}
+
+VirtualMemory::~VirtualMemory()
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/vmem.cc:73]: Good point.

File: src/main.cc (hunk 2 above)
Comment (sethpugsley) [reply src/main.cc:573]: Everything about ChampSim is deterministic.  I don't want any part of ChampSim to ever deviate from that.

File: src/vmem.cc [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  if(page_table[cpu_num][level].find(pt_lookup_tag) == page_table[cpu_num][level].end())
+    {
+      // this PTE doesn't yet have a mapping
+      page_table[cpu_num][level][pt_lookup_tag] = get_next_free_ppage();
+    }
+  
+  return (((page_table[cpu_num][level][pt_lookup_tag])<<log2_page_size)+(pte_offset*8));
+}
+
+uint64_t VirtualMemory::vmem_rand()
```
[the remark this answers is not in the document]
Comment (sethpugsley) [reply src/vmem.cc:127]: I don't mind someone else changing this to use a different random number generator in the future, as long as it maintains ChampSim's dedication to determinism, and has 0 possibility of affecting random numbers generated elsewhere in the simulation.  The current solution accomplishes both of these goals.

File: inc/vmem.h (hunk 1 above)
Comment (sethpugsley) [reply inc/vmem.h:8]: It's a number that came from mashing my keyboard.  It's used in the algorithm just to make sure that we don't get stuck in a loop of the RNG always producing 0s.

I tested the distribution of numbers generated by this RNG in a separate program, and I found it to be acceptable.

**Review by ngober — APPROVED** [review]
I think this is a good start. I particularly like the interface to get page table page numbers. That will be very useful later. I disagree with some design choices, particularly the array of map pointers, but I'm not going to cause a huge fuss if it works.
